cassovary-core/src/main/scala/com/twitter/cassovary/graph/DynamicDirectedGraphUnion.scala

```diff
@@ -0,0 +1,73 @@
+package com.twitter.cassovary.graph
+
+import com.twitter.cassovary.graph.StoredGraphDir.StoredGraphDir
+import com.twitter.cassovary.graph.node.DynamicNode
+
+/**
+ * Wraps a directed graph and a dynamic directed graph to create a dynamic graph with the union of their nodes and edges.  When
+ * edge are added, they are added to the underlying dynamic graph.  Node or edge deletion is not supported.
+ */
+class DynamicDirectedGraphUnion(staticGraph: DirectedGraph[Node], dynamicGraph: DynamicDirectedGraph[DynamicNode])
+    extends DynamicDirectedGraph[DynamicNode] {
+  // Because computing nodeCount as an intersection is expensive, maintain nodeCount as a variable.
+  private var _nodeCount = if (dynamicGraph.nodeCount == 0)
+      staticGraph.nodeCount
+    else
+    ((staticGraph map (_.id)).toSet ++ (dynamicGraph map (_.id)).toSet).size
+
+  override def getNodeById(id: Int): Option[DynamicNode] =
+    if (staticGraph.existsNodeId(id) || dynamicGraph.existsNodeId(id)) {
+      Some(new DynamicNodeUnion(
+        staticGraph.getNodeById(id),
+        dynamicGraph.getOrCreateNode(id)))
+    } else {
+      None
+    }
+
+  override def getOrCreateNode(id: Int): DynamicNode = {
+    if (!staticGraph.existsNodeId(id) && !dynamicGraph.existsNodeId(id)) {
+      _nodeCount += 1
+    }
+    new DynamicNodeUnion(staticGraph.getNodeById(id), dynamicGraph.getOrCreateNode(id))
+  }
+
+  /**
+   * Adds the given edge to the underlying dynamic graph. Note that for efficiency we don't check if the edge already exists,
+   * so if the edge already exists, a 2nd copy of it will be added.
+   */
+  override def addEdge(srcId: Int, destId: Int): Unit = dynamicGraph.addEdge(srcId, destId)
+
+  /** Not supported. */
+  override def removeEdge(srcId: Int, destId: Int): (Option[Node], Option[Node]) = throw new UnsupportedOperationException()
+
+  override def edgeCount: Long = staticGraph.edgeCount + dynamicGraph.edgeCount
+  
+  override def nodeCount: Int = _nodeCount
+
+  assert(staticGraph.storedGraphDir == dynamicGraph.storedGraphDir)
+  override val storedGraphDir: StoredGraphDir = dynamicGraph.storedGraphDir
+
+  override def iterator: Iterator[DynamicNode] = {
+    val staticGraphIds = staticGraph.iterator map (_.id)
+    val additionalDynamicGraphIds = dynamicGraph.iterator map (_.id) filter (!staticGraph.existsNodeId(_))
+    (staticGraphIds ++ additionalDynamicGraphIds) map (id => getNodeById(id).get)
+  }
+}
+
+/** Represents the union of two nodes. */
+private class DynamicNodeUnion(staticNodeOption: Option[Node],
+                               dynamicNode: DynamicNode) extends DynamicNode {
+  override val id: Int = dynamicNode.id
+
+  override def inboundNodes(): Seq[Int] =
```

Comment:
Good point!  It looks like the `++` operator is actually creating a new Seq and copying all the elements, which would be slow.  I'll add `IndexedSeqUnion` (which looks like it will require modifying `Node` to return an `IndexedSeqUnion[Int`] for neighbors, rather than `Seq[Int]`, but that seems like the proper type in general to me anyway.